src/python/pants/java/nailgun_protocol.py

```diff
@@ -232,7 +234,32 @@ def send_exit(cls, sock, payload=b''):
   @classmethod
   def send_pid(cls, sock, pid):
     """Send the PID chunk over the specified socket."""
-    cls.write_chunk(sock, ChunkType.PID, pid)
+    encoded_pid = cls.encode_int(pid)
+    cls.write_chunk(sock, ChunkType.PID, encoded_pid)
+
+  @classmethod
+  def encode_int(cls, obj):
+    """Verify the object is an int, and encode it to the right type of string for nailgun.
+
+    :param int obj: An integer to be encoded.
+    :raises: :class:`TypeError` if `obj` is not an integer.
+    :return: A representation of the int `obj` suitable to pass as the `payload` to send_exit().
+    """
+    if not isinstance(obj, int):
+      raise TypeError("cannot encode non-integer object in encode_int(): object was {} (type '{}')."
+                      .format(obj, type(obj)))
+    return binary_type(str(obj).encode('ascii'))
```

Comment:
`binary_type()` is unnecessary, because `str()` will get unicode and `encode()` will get bytes.

Reply:
Totally right! I misinterpreted a python 3 test failure in the exact opposite way I should have, leading to very helpful errors in both shards.